app/redux/Actions.js

```diff
@@ -0,0 +1,51 @@
+import apiClient from "../utils/ApiClient";
+import {HIDE_PROGRESS_DIALOG, SHOW_PROGRESS_DIALOG, USER_LOGS_IN} from "./Types";
+import store from "./Store";
+import showPopup from "../components/Popup";
+
+export function loginUser(username, password) {
+  return (dispatch) => {
+    const url = "TODO";
+    const request = {
+      username: username,
+      password: password
+    };
+    showProgressBar();
+    apiClient.post(url, request)
+      .then(response => {
+        /*
+        FIXME:
+         1. The `response` object will contain data like `userid`, `username` and `authtoken`.
+        */
```

Comment:
I was looking at the login API implementation on the server end and found that the login credentials were being saved into the session object. I don't think that will work here, so we might have to make alterations to the backend implementation.

Reply:
Agreed. The React app authentication only works because the React app is part of the core application so it has access to the same browser session as users within the Grails web app. 

But you are correct we should implement the authentication properly now and even move the React app to use this new mechanism. We obviously still need to pass the credentials to an authentication service, but how does the flow normally work with a React Native app? 

Here's the general flow (https://reactnavigation.org/docs/auth-flow/) which it seems you've implemented. A very simple authentication / authorization mechanism would probably look like this:

1. User sends credentials via Basic Auth
2. Server check credentials 
3a. On failure, server sends back 401
3b. On success, server sends back a JWT that the app stores locally 
4. On subsequent requests, server adds JWT as a Bearer token to Authorization header 

Some examples
* https://jasonwatmore.com/post/2018/09/11/react-basic-http-authentication-tutorial-example
* https://www.digitalocean.com/community/tutorials/how-to-add-login-authentication-to-react-applications

Another option would be to delegate to an Authentication service like Okta or Auth0. Since the actual authentication is currently done by OpenBoxes (and not a social provider like Google), I'd prefer not to have to implement our own OAuth provider if we don't need to (unless it's now way easier than I remember).

Reply:
> But you are correct we should implement the authentication properly now and even move the React app to use this new mechanism

We could do it without having to touch the existing React app. The mobile app and the web-app could send identifiers which the server could use to decide how the authentication should be handled. In case of the mobile app client, it would send back the details of the authenticated user(including the JWT) whereas in case of the web-app client it would continue to work the same way it works now.    

> On subsequent requests, server adds JWT as a Bearer token to Authorization header

I think you meant **client** adds JWT in the header of every subsequent request. But yeah this is exactly what we want to do. 

> Since the actual authentication is currently done by OpenBoxes (and not a social provider like Google), I'd prefer not to have to implement our own OAuth provider if we don't need to (unless it's now way easier than I remember).

I have never used something like Auth0 or Okta but I imagine they would be easy enough to integrate. These services are not free so I imagine that is something that needs to be considered as well. 

But starting with exchanging an expirable JWT on login and then expecting that token to be available for every authenticated request shouldn't be that massive of an undertaking.

Reply:
> I think you meant client adds JWT in the header of every subsequent request.

Yes, I meant client. Good catch. 

> I have never used something like Auth0 or Okta but I imagine they would be easy enough to integrate. These services are not free so I imagine that is something that needs to be considered as well.

Regarding Auth0 and Okta the cost was exactly what I was worried about as well. It would be great to provide the ability to authenticate against an external auth provider, but we need to have a way for implementers to do it on their own for free as well.

> But starting with exchanging an expirable JWT on login and then expecting that token to be available for every authenticated request shouldn't be that massive of an undertaking.

Sounds like a plan.